node/src/components/small_network/limiter.rs

```diff
@@ -178,165 +202,111 @@ impl Limiter for ClassBasedLimiter {
-                            // Determine time delta since last refill.
-                            let now = Instant::now();
-                            let elapsed = now - last_refill;
-                            last_refill = now;
-
-                            // Add appropriate amount of resources, capped at `max_stored_bytes`.
-                            resources_available +=
-                                ((elapsed.as_nanos() * resources_per_second as u128)
-                                    / 1_000_000_000) as i64;
-                            resources_available =
-                                resources_available.min(max_stored_resource as i64);
-
-                            // If we do not have enough resources available, sleep until we do.
-                            if resources_available < 0 {
-                                let estimated_time_remaining = Duration::from_millis(
-                                    (-resources_available) as u64 * 1000
-                                        / resources_per_second as u64,
-                                );
-
-                                tokio::time::sleep(estimated_time_remaining).await;
-                                wait_time_sec.inc_by(estimated_time_remaining.as_secs_f64());
-                            }
+        match peer_class {
+            PeerClass::ActiveValidator | PeerClass::UpcomingValidator => {
+                // No limit imposed on validators.
+                return;
+            }
+            PeerClass::Bulk => {
+                let max_stored_resource = ((self.data.resources_per_second as f64)
+                    * STORED_BUFFER_SECS.as_secs_f64())
+                    as u32;
+
+                // We are a low-priority sender. Obtain a lock on the resources and wait an
+                // appropriate amount of time to fill them up.
+                {
+                    let mut resources = self.data.resources.lock().await;
+
+                    while resources.available < 0 {
+                        // Determine time delta since last refill.
+                        let now = Instant::now();
+                        let elapsed = now - resources.last_refill;
+                        resources.last_refill = now;
+
+                        // Add appropriate amount of resources, capped at `max_stored_bytes`. We
+                        // are still maintaining the lock here to avoid issues with other
+                        // low-priority requestors.
+                        resources.available += ((elapsed.as_nanos()
+                            * self.data.resources_per_second as u128)
+                            / 1_000_000_000) as i64;
+                        resources.available = resources.available.min(max_stored_resource as i64);
+
+                        // If we do not have enough resources available, sleep until we do.
+                        if resources.available < 0 {
```

Comment:
How can this be negative if we've just added at least `elapsed / 1_000_000_000` to it?